subxt/src/backend/mod.rs

```diff
@@ -1125,140 +925,46 @@ mod test {
-                        Ok(operation_continue("Id1")),
-                    ])),
-                ),
-                (
-                    "chainHead_v1_storage",
-                    Message::Many(Ok(vec![
-                        Ok(storage_items("Id1", &[storage_result("ID2", "Data2")])),
-                        Ok(operation_continue("Id1")),
-                    ])),
-                ),
-                (
-                    "chainHead_v1_storage",
-                    Message::Many(Ok(vec![
-                        Ok(storage_items("Id1", &[storage_result("ID3", "Data3")])),
-                        Ok(storage_done("Id1")),
-                    ])),
-                ),
-            ];
-            let rpc_client = setup_mock_rpc_client(false)
-                .add_method("chainHead_v1_storage", |data, sub, _| {
-                    Box::pin(async move {
-                        let response = data.pop("method_response".into()).unwrap_single();
-                        if response.is_ok() {
-                            let item = data.pop("chainHead_v1_storage".into());
-                            if let Some(sub) = sub {
-                                let item = item;
-                                sub.write_delayed(item).await
-                            }
-                        }
-                        response.map(|x| RawValue::from_string(x).unwrap())
-                    })
+            let rpc_client = mock_client_builder(rx)
+                .method_handler_once("chainHead_v1_storage", move |_params| async move {
+                    // First call; return DisconnectedWillReconnect
+                    Err::<Infallible, _>(subxt_rpcs::Error::DisconnectedWillReconnect("..".into()))
                 })
-                .add_method("chainHead_v1_continue", |data, sub, _| {
-                    Box::pin(async move {
-                        let response = data.pop("continue_response".into()).unwrap_single();
-                        if response.is_ok() {
-                            let item = data.pop("chainHead_v1_storage".into());
-                            if let Some(sub) = sub {
-                                let item = item;
-                                sub.write_delayed(item).await
-                            }
-                        }
-                        response.map(|x| RawValue::from_string(x).unwrap())
-                    })
+                .method_handler_once("chainHead_v1_storage", move |_params| async move {
+                    // Next call, return a storage item and then a "waiting for continue".
+                    tokio::spawn(async move {
+                        tx.send(storage_items("Id1", &[storage_result("ID1", "Data1")]))
+                            .unwrap();
+                        tx.send(operation_continue("Id1")).unwrap();
+                    });
+                    Ok(Json(response_started("Id1")))
```

Comment:
Technically, we should ensure that response_started was sent "before" the messages in the tokio::spawn above.

if the tests passes it's not a big deal but you use a oneshot for that here.

Reply:
Actually, I don't mind if some data comes back _before_ the "response started" because, owing to races, there's a chance that could happen in real life too, so this checks that the client is ready to receive such messages before the response comes back (which I think we do; we subscribe before making the call to buffer any responses)